### PR TITLE
[Runtime] Enhance PackedFunc Metaprogramming with `PackArgs`

### PR DESCRIPTION
**Pull request by junrushao**
This PR introduces a convenient method `PackArgs` that allows packing variadic C++ parameters into TVM's PackedFunc calling convention without recursion using C++ parameter pack tricks.

As an example,

```C++
template <typename... Args>
void MyVariadicMethod(Args&&... args) {
  constexpr int kNumArgs = sizeof...(Args);
  TVMValue values[kNumArgs];
  int type_codes[kNumArgs];
  tvm::runtime::PackArgs(values, type_codes, std::forward<Args>(args)...);
  ...
}
```

### CODE REVIEW

**Comment by junrushao** [comment]
CC: @cyx-6 @tqchen

**Review by cyx-6 — APPROVED** [review]
LGTM!

File: include/tvm/runtime/packed_func.h [hunk 1]
```diff
@@ -1622,6 +1641,20 @@ inline TVMRetValue PackedFunc::operator()(Args&&... args) const {
   return rv;
 }
 
+template <int i, typename T>
```
Comment (kparzysz-quic) [inline include/tvm/runtime/packed_func.h:1644]: This should be `size_t` not `int`.  Otherwise it fails to compile with clang.

Comment (junrushao) [reply include/tvm/runtime/packed_func.h:1644]: Yeah I found the same thing: https://github.com/apache/tvm/pull/15637...It works with my GCC but not clang

**Pull request merged**